### DISCUSSION TITLE
How to translate this piece of code to PascalABC.NET?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#2946] [General]
```
{$apptype windows}

uses
  Eto.Forms,
  Eto.Drawing,
  Eto;

type
  MyObject = class
  private
    textProperty: string;
    procedure setText(value: string);
    begin
      textProperty := value;
      System.Diagnostics.Debug.WriteLine(string.Format('Set TextProperty to {0}', value));
    end;
  
  public
    property Text: string read textProperty write setText;
  end;
  
  MyForm = class(Form)
  public
    constructor;
    begin
      Title := 'Table Layout';
      ClientSize := new Eto.Drawing.Size(600, 400);
      
      var tblLayout := new TableLayout;
      tblLayout.Spacing := new Eto.Drawing.Size(5, 5);
      tblLayout.Padding := new Eto.Drawing.Padding(10);
      
      var lbl1 := new &Label;
      lbl1.Text := 'DataContext Binding';
      var tblRow1 := new TableRow;
      tblRow1.Cells.Add(new TableCell(lbl1, true));
      tblRow1.Cells.Add(new TableCell(DataContextBinding, true));
      
      var lbl2 := new &Label;
      lbl2.Text := 'Object Binding';
      var tblRow2 := new TableRow;
      tblRow2.Cells.Add(new TableCell(lbl2, true));
      tblRow2.Cells.Add(new TableCell(ObjectBinding, true));
      
      var lbl3 := new &Label;
      lbl3.Text := 'Direct Binding';
      var tblRow3 := new TableRow;
      tblRow3.Cells.Add(new TableCell(lbl3, true));
      tblRow3.Cells.Add(new TableCell(DirectBinding, true));
      
      tblLayout.Rows.Add(tblRow1);
      tblLayout.Rows.Add(tblRow2);
      tblLayout.Rows.Add(tblRow3);
      tblLayout.Rows.Add(nil);
      
      Content := tblLayout;
      
      var myObj := new MyObject;
      myObj.Text := 'Initial value 1';
      
      DataContext := myObj;
      
      Menu := new Eto.Forms.MenuBar;
      Menu.QuitItem := new Command((sender, e) -> Application.Instance.Quit());
      Menu.QuitItem.Text := 'Quit';
      Menu.QuitItem.Shortcut := Application.Instance.CommonModifier or Keys.Q;
    end;
    
    function DataContextBinding: TextBox;
    begin
      var textBox := new TextBox();
      //Project4.pas(72) : Found '<' but expected ';'
      //textBox.TextBinding.BindDataContext<MyObject>(r -> r.Text, (r, val) -> r.Text = val);
      result := textBox;
    end;
    
    function ObjectBinding: TextBox;
    begin
      var obj := new MyObject;
      obj.Text := 'Initial value 2';
      var textBox := new TextBox();
      //Project4.pas(81) : No overloaded function Bind with same arguments
      //textBox.TextBinding.Bind(obj, r -> r.Text);
      result := textBox;
    end;
    
    function DirectBinding: TextBox;
    begin
      var textBox := new TextBox();
      textBox.TextBinding.Bind(() -> 'some value', val -> System.Diagnostics.Debug.WriteLine(string.Format('Set value to {0} directly', val)));
      result := textBox;
    end;
  end;

begin
  (new Application).Run(new MyForm);
end.
```

Note: the problematic lines are commented out. Original C# source:

https://github.com/picoe/Eto/blob/develop/samples/Tutorials/CSharp/Tutorial4/Main.cs

Eto source code could be referenced directly here:

https://github.com/picoe/Eto/blob/develop/src/Eto/Forms/Controls

https://github.com/picoe/Eto/tree/develop/src/Eto/Forms/Binding

The results of `Go To Declaration` and `ILSpy` are too complex. This is why I decided to give the link to Eto source code directly.

It's not as simple as translating the C# lambdas syntax to PascalABC.NET. I tried and failed.

**Comment by samuraiGH** [comment]
`.BindDataContext<MyObject>` -> `.BindDataContext&<MyObject>`

**Reply by SunSerega** [reply]
The setter is also untranslated:
```
      textBox.TextBinding.BindDataContext&<MyObject>(r -> r.Text, (r, val) -> (r.Text := val));
```

**Comment by samuraiGH** [comment]
`.Bind(obj, r -> r.Text);`

in example:
`.Bind(obj, r => r.TextProperty);`

maybe
`.Bind(obj, r -> r.TextProperty);`

**Reply by SunSerega** [reply]
`.TextProperty` is defined at the beginning of Pascal code, TS changed the name from the example when defining it.

**Comment by SunSerega** [comment]
> ```
>       //Project4.pas(81) : No overloaded function Bind with same arguments
>       //textBox.TextBinding.Bind(obj, r -> r.Text);
> ```

Right... https://github.com/picoe/Eto/blob/develop/src/Eto/Forms/Binding/ObjectBinding.cs#L380
You've been hitting all the missing syntax sugar in Pascal :)

This overload makes the C# compiler generate an `Expression` object, instead of an actual lambda function. From this `Expression` object the library code can extract what property you have returned and automatically generate a setter lambda.
But the Pascal compiler doesn't know how to generate these `Expression` objects.

You can create them manually, but then it misses the point of writing less.
Well, you can still just specify getter and setter lambdas, like you already did for other bindings.

**Reply by ghost** [reply]
@SunSerega Sorry for my very late reply. This has been so long and I have forgot a lot of things. When I'm reading old code written by me now, I have no idea how it's working. Please give me complete code that works. Thank you.

**Reply by SunSerega** [reply]
All that applies to me even more. I never properly figured out what's happening or even used the library, only answering per-case and compiling code in my head.

Besides, it is one of the ways to strengthen knowledge - reconstructing the understanding from what little you do remember. Each time it's easier because the bigger portion is known intuitively.

**Reply by ghost** [reply]
I feel the PascalABC.NET language is desperately lacking a lot of modern features (that you called syntax sugars), making mapping modern C# code to PascalABC.NET code is increasingly difficult. I will give up on using `Eto.Forms` with PascalABC.NET.

**Reply by SunSerega** [reply]
Well, I agree that `Expression`-from-lambda generation is a vital missing feature, but your experience with the critical lack of modern features is a mix of bad luck and giving up too early.

In the first place, "syntax sugar" is not a general term for modern features. It's what you add on top of a language to make it sweeter (easier) to use (in this case during the syntax stage of compiling).
In other words "sugar" features are those replaceable using more basic features. Doesn't even have to be modern - like how `break` is replaceable by `goto`, but is generally considered better.

Here, I already told you about another overload that takes in getter and setter lambdas. And there are even more overloads to simplify other cases, like adding property by name.
So it seems to me the authors of that library did care about making everything accessible in a very generic use case.
But I'm not going to do any more work of chewing info for you in this direction when you are talking about giving up.
Both because it's self-destructive on your side, and because this kind of talk is unpleasant to me.

**Reply by ghost** [reply]
@SunSerega I don't want to give up, too. This is my project. Please help.

[Project4.zip](https://github.com/pascalabcnet/pascalabcnet/files/13943942/Project4.zip)

**Reply by SunSerega** [reply]
Got the `DataContextBinding` function working like this:
```
      var text_get := function(r: MyObject): string -> r.Text;
      var text_set := procedure(r: MyObject; val: string) -> (r.Text := val);
      textBox.TextBinding.BindDataContext(text_get, text_set);
```
At least the method call works now, I'll still need to figure out what is wrong with the compiler - probably got confused by all the overloads.

But I didn't dig in enough into `Eto` to understand how this binding is supposed to show up after launch.

**Reply by SunSerega** [reply]
As for `ObjectBinding`, here is how you create expressions manually:
```
      var par := System.Linq.Expressions.Expression.Parameter(typeof(MyObject), 'r');
      var text_prop_get := System.Linq.Expressions.Expression.Property(par, 'Text');
      textBox.TextBinding.Bind(obj, System.Linq.Expressions.Expression.Lambda&<MyObject->string>(text_prop_get, par));
```
(basically copied everything with minimal translation from Phind)

But this is pretty pointless. The whole point of those expressions is to write less, letting the compiler generate everything.
At this point, you are better off using overloads that let you specify the property name:
```
      textBox.TextBinding.BindDataContext&<MyObject>('Text');
```
As I understand, the library does the same thing as above, manually creating the' Expression' object using the property name you've passed in. Except it also creates the setter for that property.

**Reply by SunSerega** [reply]
> I'll still need to figure out what is wrong with the compiler

https://github.com/pascalabcnet/pascalabcnet/issues/3029
